File: internal/users/session_test.go
package users

import (
	"errors"
	"strings"
	"testing"
	"time"
)

const sessionKey = "a-signing-key-for-tests-only"

func TestASessionVerifiesAndNamesItsUser(t *testing.T) {
	s := NewSessions(sessionKey)
	now := time.Now()

	got, err := s.Verify(s.Issue("user-1", now), now)
	if err != nil {
		t.Fatalf("Verify: %v", err)
	}
	if got != "user-1" {
		t.Errorf("user = %q, want user-1", got)
	}
}

// The expiry lives INSIDE the signature, so extending a session means forging
// one rather than editing a field. A client holds this value, and a client is
// not a thing to take an expiry claim from.
func TestAnExtendedExpiryDoesNotVerify(t *testing.T) {
	s := NewSessions(sessionKey)
	now := time.Now()
	token := s.Issue("user-1", now)

	// Rewrite the expiry to the far future, keeping the signature.
	rest := strings.TrimPrefix(token, "usr_")
	cut := strings.LastIndex(rest, ".")
	signature := rest[cut+1:]
	forged := "usr_user-1.99999999999." + signature

	if _, err := s.Verify(forged, now); !errors.Is(err, ErrInvalidSession) {
		t.Errorf("a rewritten expiry verified: %v", err)
	}
}

func TestATamperedSignatureDoesNotVerify(t *testing.T) {
	s := NewSessions(sessionKey)
	now := time.Now()
	token := s.Issue("user-1", now)

	flipped := token[:len(token)-1] + map[bool]string{true: "b", false: "a"}[strings.HasSuffix(token, "a")]
	if _, err := s.Verify(flipped, now); !errors.Is(err, ErrInvalidSession) {
		t.Errorf("a tampered signature verified: %v", err)
	}
}

func TestASessionExpires(t *testing.T) {
	s := NewSessions(sessionKey)
	now := time.Now()
	token := s.Issue("user-1", now)

	if _, err := s.Verify(token, now.Add(SessionTTL-time.Minute)); err != nil {
		t.Errorf("a session expired early: %v", err)
	}
	if _, err := s.Verify(token, now.Add(SessionTTL+time.Minute)); !errors.Is(err, ErrInvalidSession) {
		t.Errorf("an expired session verified: %v", err)
	}
}

// A session signed with one key must not verify under another. This is what
// makes the per-process fallback key safe: a restart mints a new key and every
// outstanding session stops working, which is the intended behaviour and only
// holds if the key participates in the signature.
func TestASessionDoesNotVerifyUnderADifferentKey(t *testing.T) {
	now := time.Now()
	token := NewSessions(sessionKey).Issue("user-1", now)

	if _, err := NewSessions("a-different-key-entirely").Verify(token, now); !errors.Is(err, ErrInvalidSession) {
		t.Errorf("a session verified under a different key: %v", err)
	}
}

// Two Sessions built with no key must not agree, or the "restart signs everyone
// out" property would be a comment rather than a behaviour.
func TestAnUnsetKeyIsGeneratedPerInstance(t *testing.T) {
	now := time.Now()
	token := NewSessions("").Issue("user-1", now)

	if _, err := NewSessions("").Verify(token, now); !errors.Is(err, ErrInvalidSession) {
		t.Error("two instances with no configured key produced the same signing key")
	}
}

func TestMalformedSessionsAreRefused(t *testing.T) {
	s := NewSessions(sessionKey)
	now := time.Now()

	for _, bad := range []string{
		"", "usr_", "usr_nodot", "usr_user-1.notanumber.abc",
		"user-1.123.abc",           // missing prefix
		"usr_.123.abc",             // empty user id
		s.Issue("user-1", now)[4:], // prefix stripped
	} {
		if _, err := s.Verify(bad, now); !errors.Is(err, ErrInvalidSession) {
			t.Errorf("Verify(%q) err = %v, want ErrInvalidSession", bad, err)
		}
	}
}

// The API routes a bearer value to the right verifier by prefix, so a
// configured token that happens to contain a dot is not run through the session
// verifier.
func TestOnlyPrefixedValuesLookLikeSessions(t *testing.T) {
	if !LooksLikeSession(NewSessions(sessionKey).Issue("user-1", time.Now())) {
		t.Error("an issued session does not look like one")
	}
	for _, notASession := range []string{
		"secureops-configured-token-value", "abc.def.ghi", "", "user_something",
	} {
		if LooksLikeSession(notASession) {
			t.Errorf("%q looks like a session", notASession)
		}
	}
}

File: apps/web/src/lib/session.test.ts
/**
 * @vitest-environment node
 *
 * No DOM here: this module is `node:crypto` and environment variables. Running
 * it under jsdom would be slower and would test it in a context it never runs
 * in.
 */
import { randomBytes } from "node:crypto";
import { beforeEach, describe, expect, it, vi } from "vitest";

/**
 * The session boundary (ADR 029, ADR 031).
 *
 * These are the invariants the login rests on, and every one of them is
 * invisible from the outside: a cookie that fails any of them looks exactly
 * like a cookie that passes, right up until the server checks it.
 *
 * The module reads its key at import time, so each test re-imports it under a
 * fresh module registry rather than sharing one instance. Otherwise the first
 * test's key would silently decide every later test's result.
 */
async function load(env: Record<string, string | undefined>) {
  vi.resetModules();
  for (const [key, value] of Object.entries(env)) {
    if (value === undefined) delete process.env[key];
    else process.env[key] = value;
  }
  return import("./session");
}

/**
 * Generated per run rather than written down.
 *
 * A literal 32-character hex string in a source file is indistinguishable from
 * a real signing key, and gitleaks flagged the first version of this file for
 * exactly that -- correctly. Allowlisting the line would have taught the
 * scanner to ignore a shape it should always question; generating the value
 * removes the shape. The tests do not depend on any particular key, only on
 * two of them differing.
 */
const key = () => randomBytes(16).toString("hex");
const KEY = key();

beforeEach(() => {
  delete process.env.SECUREOPS_DASHBOARD_PASSWORD;
  delete process.env.SECUREOPS_DASHBOARD_SESSION_KEY;
});

describe("passwordMatches", () => {
  it("accepts the configured password", async () => {
    const { passwordMatches } = await load({ SECUREOPS_DASHBOARD_PASSWORD: "correct horse" });
    expect(passwordMatches("correct horse")).toBe(true);
  });

  it("rejects a wrong password", async () => {
    const { passwordMatches } = await load({ SECUREOPS_DASHBOARD_PASSWORD: "correct horse" });
    expect(passwordMatches("correct hors")).toBe(false);
    expect(passwordMatches("correct horse ")).toBe(false);
    expect(passwordMatches("")).toBe(false);
  });

  // The permissive default ADR 006 refuses. An unset password must lock the
  // door, not remove it -- and the failure is silent either way, so it is
  // asserted rather than assumed.
  it("rejects everything when no password is configured", async () => {
    const { passwordMatches, dashboardPasswordConfigured } = await load({
      SECUREOPS_DASHBOARD_PASSWORD: undefined,
    });
    expect(dashboardPasswordConfigured()).toBe(false);
    expect(passwordMatches("")).toBe(false);
    expect(passwordMatches("anything")).toBe(false);
  });

  it("treats a whitespace-only password as unconfigured", async () => {
    const { passwordMatches, dashboardPasswordConfigured } = await load({
      SECUREOPS_DASHBOARD_PASSWORD: "   ",
    });
    expect(dashboardPasswordConfigured()).toBe(false);
    expect(passwordMatches("   ")).toBe(false);
  });
});

describe("sessionIsValid", () => {
  it("accepts a session it just issued", async () => {
    const { issueSession, sessionIsValid } = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    expect(sessionIsValid(issueSession())).toBe(true);
  });

  it("rejects a tampered signature", async () => {
    const { issueSession, sessionIsValid } = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    const [expires, signature] = issueSession().split(".");
    const flipped = signature!.startsWith("a")
      ? `b${signature!.slice(1)}`
      : `a${signature!.slice(1)}`;

    expect(sessionIsValid(`${expires}.${flipped}`)).toBe(false);
  });

  /**
   * The property that makes carrying the expiry inside the signature worth
   * doing, and the only place it is asserted.
   *
   * A cookie's own Max-Age is a claim the browser makes, and a browser is not a
   * thing to take an expiry claim from. If the expiry were outside the
   * signature, this rewrite would grant an indefinite session.
   */
  it("rejects an expiry rewritten to the far future", async () => {
    const { issueSession, sessionIsValid } = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    const signature = issueSession().split(".")[1];

    expect(sessionIsValid(`99999999999999.${signature}`)).toBe(false);
  });

  it("rejects a session past its expiry", async () => {
    const { issueSession, sessionIsValid } = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    const now = Date.now();
    const value = issueSession(now);

    expect(sessionIsValid(value, now + 60_000)).toBe(true);
    // Eight hours and a minute later.
    expect(sessionIsValid(value, now + 8 * 60 * 60 * 1000 + 60_000)).toBe(false);
  });

  it("rejects malformed and absent values", async () => {
    const { sessionIsValid } = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    for (const value of [undefined, "", ".", "nodot", "abc.", ".abc", "not-a-number.abc"]) {
      expect(sessionIsValid(value)).toBe(false);
    }
  });

  /**
   * A session signed with one key must not validate under another.
   *
   * This is what makes the per-process fallback key safe: with no
   * SECUREOPS_DASHBOARD_SESSION_KEY set, a restart mints a new key and every
   * outstanding session stops working. That is the intended behaviour, and it
   * only holds if the key actually participates in the signature.
   */
  it("rejects a session signed with a different key", async () => {
    const first = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: KEY });
    const value = first.issueSession();

    const second = await load({ SECUREOPS_DASHBOARD_SESSION_KEY: key() });
    expect(second.sessionIsValid(value)).toBe(false);
  });
});
